bombe/src/main/java/org/cadixdev/bombe/jar/Jars2.java

```diff
@@ -0,0 +1,148 @@
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.jar.Manifest;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Utilities for working with jar files.
+ *
+ * @author Jamie Mansfield
+ * @since 0.4.0
+ */
+public final class Jars2 {
+
+    public static Stream<AbstractJarEntry> walk(final FileSystem fs) throws IOException {
+        final Path root = fs.getPath("/");
+
+        final List<AbstractJarEntry> entries = new ArrayList<>();
+
+        Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFile(final Path file, final BasicFileAttributes attrs) throws IOException {
+                entries.add(read(file));
+                return FileVisitResult.CONTINUE;
+            }
+        });
+
+        return entries.stream();
+    }
+
+    /**
+     * Creates an {@link AbstractJarEntry} for a given path.
+     *
+     * @param entry The path of the jar entry
+     * @return The jar entry
+     * @throws IOException Should the entry fail to read
+     */
+    public static AbstractJarEntry read(final Path entry) throws IOException {
+        try (final InputStream is = Files.newInputStream(entry)) {
+            final String name = entry.toString().substring(1); // Remove '/' prefix
+            final long time = Files.getLastModifiedTime(entry).toMillis();
+
+            if (Objects.equals("META-INF/MANIFEST.MF", name)) {
+                return new JarManifestEntry(time, new Manifest(is));
+            }
+            else if (name.startsWith("META-INF/services/")) {
+                final String serviceName = name.substring("META-INF/services/".length());
+
+                final ServiceProviderConfiguration config = new ServiceProviderConfiguration(serviceName);
+                config.read(is);
+                return new JarServiceProviderConfigurationEntry(time, config);
+            }
+
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            ByteStreams.copy(is, baos);
+
+            if (name.endsWith(".class")) {
+                return new JarClassEntry(name, time, baos.toByteArray());
+            }
+            else {
+                return new JarResourceEntry(name, time, baos.toByteArray());
+            }
+        }
+    }
+
+    public static void transform(final FileSystem input, final Path output, final JarEntryTransformer... transformers) throws IOException {
+        final URI uri = URI.create("jar:" + output.toUri());
+        final Map<String, String> options = new HashMap<String, String>() {
+            {
```

Comment:
Creating an inner class just to initialize a `HashMap` is ugly imo, just put this outside after creating it.